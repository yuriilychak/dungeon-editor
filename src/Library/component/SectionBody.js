import React from "react";
import {makeStyles} from "@material-ui/styles";
import SortableTree from "react-sortable-tree";
import FileExplorerTheme from "react-sortable-tree-theme-file-explorer";

import FolderClosedIcon from "@material-ui/icons/Folder";
import FolderOpenIcon from "@material-ui/icons/FolderOpen";
import DeleteIcon from '@material-ui/icons/Delete';
import EditIcon from '@material-ui/icons/Edit';
import FolderAdd from '@material-ui/icons/CreateNewFolder';

import ToolButton from "./ToolButton";
import Icon from "./Icon";
import "../styles/expand-menu.css";

const useStyles = makeStyles({
    iconStyle: {
        fontSize: 16,
        color: "white"
    },
    sectionText: {
        color: "white",
        fontSize: 14
    },
    emptySection: {
        padding: "0 12px 6px 12px"
    }
});

const SectionBody = ({
                         id,
                         files,
                         icon,
                         emptyText,
                         renameText,
                         deleteText,
                         addDirectoryText,
                         onAddDirectory,
                         onRemoveFile,
                         onUpdateTree
                     }) => {
    const {iconStyle, sectionText, emptySection} = useStyles();

    if (files.length === 0) {
        return (
            <span className={emptySection}>
                {emptyText}
            </span>
        );
    }

    const generateNodeProps = rowInfo => {
        const isDirectory = rowInfo.node.isDirectory;
        const icons = [];
        const buttons = [
            <ToolButton
                title={renameText}
                Icon={EditIcon}
                onClick={() => {
                }}
            />,
            <ToolButton
                title={deleteText}
                Icon={DeleteIcon}
                onClick={() => onRemoveFile(rowInfo.node.id, id, isDirectory)}
            />
        ];

        if (isDirectory) {
            buttons.unshift(
                <ToolButton
                    title={addDirectoryText}
                    Icon={FolderAdd}
                    onClick={() => onAddDirectory(id, rowInfo)}
                />
            );

            icons.push(
                rowInfo.node.expanded ?
                    <FolderOpenIcon className={iconStyle}/> :
                    <FolderClosedIcon className={iconStyle}/>
            );
        } else {
            icons.push(
                <Icon name={`${icon}_element`}/>
            );
        }

        return {
            onClick: event => console.log(event),
            icons,
            buttons,
            className: sectionText
        };
    };

    let nodeCount = 0;

    const calculateHeight = elements => {
        let result = 0;
        elements.forEach(element => {
            result += 24;
            ++nodeCount;
            if (element.expanded && element.children && element.children.length !== 0) {
                result += calculateHeight(element.children);
            }
        });
        return result;
    };

    const height = calculateHeight(files) + nodeCount;

    return (
        <div style={{height, width: "100%"}}>
            <SortableTree
                treeData={files}
                onChange={fileTree => onUpdateTree(fileTree, id)}
                theme={FileExplorerTheme}
                canDrag={({node}) => !node.dragDisabled}
                canDrop={({nextParent}) => !nextParent || nextParent.isDirectory}
                scaffoldBlockPxWidth={10}
                generateNodeProps={generateNodeProps}
            />
        </div>
    );
};

export default SectionBody;
