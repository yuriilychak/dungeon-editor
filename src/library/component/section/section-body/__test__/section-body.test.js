import React from "react";
import { createMount } from "@material-ui/core/test-utils";

import SectionBody from "../section-body";

jest.mock("../../../../../common-ui/file-tree/file-tree");

describe("SectionBody test",() => {
    const defaultProps = {
        id: 0,
        icon: "test",
        emptyText: "test",
        renameText: "test",
        deleteText: "test",
        addDirectoryText: "test",
        onAddDirectory: jest.fn(),
        onRemoveFile: jest.fn(),
        onRenameFile: jest.fn(),
        onSelectFile: jest.fn(),
        onUpdateTree: jest.fn(),
        onOpenFile: jest.fn()
    };

    it("default snapshot", () => {
        const files = [
            { title: '.gitignore' },
            { title: 'package.json' },
            {
                title: 'src',
                isDirectory: true,
                expanded: true,
                children: [
                    { title: 'styles.css' },
                    { title: 'index.js' },
                    { title: 'reducers.js' },
                    { title: 'actions.js' },
                    { title: 'utils.js' },
                ],
            },
            {
                title: 'tmp',
                isDirectory: true,
                children: [
                    { title: '12214124-log' },
                    { title: 'drag-disabled-file', dragDisabled: true },
                ],
            },
            {
                title: 'build',
                isDirectory: true,
                children: [{ title: 'react-sortable-tree.js', preview: 'test' }],
            },
            {
                title: 'public',
                isDirectory: true,
            },
            {
                title: 'node_modules',
                isDirectory: true,
            },
        ];

        const wrapper = createMount()(
            <SectionBody
                {...defaultProps}
                files={files}
            />
        );

        expect(wrapper.html()).toMatchSnapshot();
    });

    it("snapshot with empty fields", () => {
        const wrapper = createMount()(
            <SectionBody
                {...defaultProps}
                files={[]}
            />
        );

        expect(wrapper.html()).toMatchSnapshot();
    });
});
