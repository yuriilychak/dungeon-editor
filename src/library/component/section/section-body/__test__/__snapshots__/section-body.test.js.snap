// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SectionBody test default snapshot 1`] = `"<div>Mock file tree</div>"`;

exports[`SectionBody test snapshot with empty fields 1`] = `"<span class=\\"section-body-empty\\">test</span>"`;
